### PR TITLE
feat: peer sync controller [1/2]

### PR DESCRIPTION
**Pull request by gmaclennan**
Hypercore manages peers on a "per-hypercore" basis, e.g. each hypercore
has a unique set of peers. However for Mapeo we need to treat all
hypercores belonging to a peer as one, so we need to combine the peer
core sync states across all cores, and then have a means for a
particular peer to control which cores / namespaces are replicated to
the replication stream.

This abstraction is for managing the enabling of namespaces with a
particular peer, and will be managed by the sync controller. This
replaces functionality that is currently in the CoreManager and the
ReplicationStateMachine that is currently returned by
coreManager.replicate().

I will include tests in a stacked pull request that implements SyncController that manages all the PeerSyncController instances.

### CODE REVIEW

**Comment by gmaclennan** [comment]
@achou11 I finally sorted this

File: src/sync/peer-sync-controller.js [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        const cap = await this.#capabilities.getCapabilities(peerId)
+        this.#syncCapability = cap.sync
+      } catch (e) {
+        // Any error, consider sync blocked
+        this.#syncCapability = createNamespaceMap('blocked')
+      }
+    }
+    // console.log(peerId.slice(0, 7), this.#syncCapability)
+    // console.log(peerId.slice(0, 7), didUpdate)
+    // console.dir(state, { depth: null, colors: true })
```
Comment (achou11) [inline src/sync/peer-sync-controller.js:130]:
```suggestion
```

**Review by achou11 — APPROVED** [review]
Couldn't find anything noteworthy to point out in the implementation. Seems good to me based on the tests you added!

File: src/sync/peer-sync-controller.js (hunk 1 above)
Comment (gmaclennan) [reply src/sync/peer-sync-controller.js:130]: I'll keep these for now for debugging integration tests.

**Pull request merged**